pagefaultgames/pokerogue
# [Enhancement][Challenge] Nuzlocke challenge

## What are the changes?

This PR adds the challenge mod "Nuzlocke" to the game. As a result, various new challenge checks will be added that can also be used for other challenges in the future.

## Why am I doing these changes?



I have seen some feature request for a nuzlocke challenge. Also, the old roadmap even had a nuzlocke mode as goal.

## What did change?



The Nuzlocke challenge aswell as the achievement for it was added. I also changed the challenge UI and made the description field larger to be able to display more text.

5 new `challengeType`'s where added to check for various rules in the nuzlocke challenge. 
Those types are:

- `NO_HEAL_PHASE`
Checks if the heal phase after every 10th round should be activated.

- `SHOP_ITEM_BLACKLIST`
Filters the shop for specific items.

- `RANDOM_ITEM_BLACKLIST`
Filters the random item selection for specific items.

- `ADD_POKEMON_TO_PARTY`
Checks if the current catched pokemon can be added to the party.

- `MOVE_BLACKLIST`
Checks if move the user wants to use is illegal for this challenge.

### Screenshots/Videos


Nuzlocke Regular:
![grafik](https://github.com/pagefaultgames/pokerogue/assets/16718523/39c0ef62-2955-4617-9b98-ba369ddecc54)

Nuzlocke Hardcore:
![grafik](https://github.com/pagefaultgames/pokerogue/assets/16718523/13692a7a-974d-43a8-b9b5-da6b77971b62)

## How to test the changes?





Set `POKEBALL_OVERRIDE` to have some masterballs and `MOVESET_OVERRIDE` to include `Moves.REVIVAL_BLESSING`.
You can also set `OPP_LEVEL_OVERRIDE` to 1.

Start a new challenge with Nuzlocke challenge enabled and check:
- if you are healed after the 10th round
- if you can buy revive
- if you can role for a reviver seed
- you can use revival blessing
- if you can only add one pokemon every 10 floors to the party


## Checklist
